### ISSUE TITLE
Issue when tring to map fields for data injection

### ISSUE THREAD

**Issue by Mehdi1962** [pluginsGLPI/genericobject#176]
Hi all,

I'm using the last versions of these 2 plugins.
Datainjection : 2.5.0
GenericObject 2.5.0
I'm under GLPI version 9.2.1

In datainjection, when I created a model and want to map filds for data injection I only have "type" and "Technician in charge of the hardware"

![image](https://user-images.githubusercontent.com/36728482/36530820-64a96f9c-17bc-11e8-8a23-c615fc2284b8.png)

Datainjection works fine with the natives models (computer, phones, etc..)

Thx a lot for your help !

**Comment by Guegui** [comment]
I have the exact same issue, any idea anyone?

**Comment by Guegui** [comment]
Hi Mehdi, have you solved your problem?

**Comment by pVesian** [comment]
It seems this occurs because there is no `linkfield` attribute in most Genericobject fields.

**Comment by chichtacos** [comment]
OK thx for the feedback ! Do you have a workaround or do you have to work on a patch ?

**Comment by Guegui** [comment]
Is that linkfield inputable somewhere in genericobjects...?

**Comment by pVesian** [comment]
@Guegui apparently it should be in `PluginGenericobjectObject::getObjectSearchOptions()`

**Comment by pVesian** [comment]
I managed to have a partial working fix. Add this to the last lines of `PluginGenericobjectObject::getObjectSearchOptions()`, just before the `$index = $currentindex + 1;`:

```php
        if ($item->canUsePluginDataInjection()) {
            // add datainjection links

            switch ($field) {
               case 'name':
               case 'template_name':
               case 'serial':
               case 'otherserial':
               case 'other':
               case 'is_helpdesk_visible':
               case 'is_recursive':
                  $options[$currentindex]['linkfield'] = $options[$currentindex]['field'];
                  break;

               default:
                  // nothing to do...
            }
         }
```

I did *not* find a way to add other fields like locations_id, states_id, groups_id, manufacturers_id, plugin_genericobject_genericobjectmodels_id, and plugin_genericobject_genericobjecttypes_id. It will have to wait until next week.

**Comment by 4n63lu5** [comment]
maybe the same problem : #193

**Comment by chichtacos** [comment]
Hi !

I'm sorry but what file do I have to edit to apply your workaround ? 

Thx !

**Comment by pVesian** [comment]
@chichtacos https://github.com/pluginsGLPI/genericobject/pull/201/commits/57f29c9916ff7ad43a3568e8990722cfd0531e89

**Comment by michel-MPH** [comment]
Works perfect. Thanks @pVesian

**Comment by pVesian** [comment]
@trasher This hasn't been merged yet, why close it?

**Comment by trasher** [comment]
Hum sorry, though this was merged... Link to commit is unclear.

**Comment by chichtacos** [comment]
Yep it works perfectly ! thx a lot @pVesian

**Comment by Andromede88** [comment]
Thank you for your work, but for me it does not work perfectly.
I don't see models, types and categories and users_id
Can you help me please ?
GLPI 9.2.1
GenericObject 2.5.2
DataInjection 2.5.2

**Comment by Andromede88** [comment]
Sorry, it's work when i re create my object, thank you very much for your work !!

**Comment by loicbenoit** [comment]
As per Cedric-Anne comment in [datainjection/issues/103](https://github.com/pluginsGLPI/datainjection/issues/103#issuecomment-415356991), this is probably related to the "linkfield of the item in search options" and it's use to populate the dropdowns.

She suggests two solutions, but both require extensive work. Basically, use something else then linkfield to populate dropdowns.

As this appears to impacts both DataInjection and GenericObject, it might be a good thing to unite our efforts and solve this issue for both plugins, if possible.

I'm willing to help, but I'm still quite new to GLPI... so I need directions.

**Comment by loicbenoit** [comment]
I'm having this same issue, where fields are missing in the dropdowns of the mapping step of DataInjection.

### List of missing fields in DataInjection dropdown for mapping a CSV file to a GenericObject:
- is_deleted
- entities_id
- serial
- Any custom field (see [GenericObject global fields](https://glpi-plugins.readthedocs.io/en/latest/genericobject/index.html#add-global-fields) and [GenericObject type field](https://glpi-plugins.readthedocs.io/en/latest/genericobject/index.html#add-new-fields))

**Not being able to import customs fields is the real issue for me.** I can do without the missing 3 standard fields as long as I'm able to import my custom fields.

### Checked
- Custom fields have been added to the object fields list (GenericObject/<type>/tab:Fields).
- GenericObject/<type>/tab:Main: Checked option to use DataInjection.
- GenericObject/<type>/tab:Main: Is active? Yes.
- GenericObject/<type>/tab:Main: Checked option "use sub-entities" (and all other options as well).
- User profile is root with all rights selected and entity is "root entity including sub-entities".

### Risk factor
- The custom GenericObject is part of a family of objects.

### Question
How do you apply this fix when method PluginGenericobjectObject::getObjectSearchOptions() was already removed to solve other issues?

See #210, #211, #212
See commit: https://github.com/pluginsGLPI/genericobject/commit/21ab730717e41f3d8cc84da994ff7a2f08ee0835#diff-c7d2c155f00fc5787c51bca2ac6a2d18

**Comment by loicbenoit** [comment]
This issue seems to impact unicity rules as well.

### How to reproduce
1. Browse to Configuration / Field Unicity (/glpi/front/fieldunicity.php)
2. Create a new rule.
3. For type, select a GenericObject.

### Expected result
Should be able to choose any field, including GLPI standard fields, as unique field.

### Current result
List is limite to 11 fields, some standard, some custom, with no obvious pattern.

### Potential cause
See probable explaination and potential solutions by Cedric-Anne:
https://github.com/pluginsGLPI/datainjection/issues/103#issuecomment-415356991

**Comment by loicbenoit** [comment]
I have proposed a fix for missing fields in unicity dropdowns. In need of reviewing...
https://github.com/pluginsGLPI/genericobject/pull/214

**Comment by loicbenoit** [comment]
Although #214 fixes unicity rule "unique" fields dropdown, it doesn't fix the issue of missing fields in the dropdowns for the mapping step of DataInjection.

**Comment by loicbenoit** [comment]
Pull-request #214 should now fix both the unicity dropdowns and field mapping in DataInjection.

### Problem analysis
GenericObjects fields were missing in DataInjection, because they would get accidentally blacklisted by other item types that were using the same indexes when declaring blacklisted fields to DataInjection.

### Proposed solution
Don't use search indexes that other item types are blacklisting in plugin DataInjection.

### Help wanted
Although the fix seems to be working well, more tests are needed to make sure it has no side effect.

**Comment by cedric-anne** [comment]
Fixed by #214 .
